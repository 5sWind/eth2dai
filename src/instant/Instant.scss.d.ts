export const panel: string;
export const header: string;
export const details: string;
export const assets: string;
export const errors: string;
export const topLeftCorner: string;
export const assetPicker: string;
export const swapIcon: string;
export const input: string;
