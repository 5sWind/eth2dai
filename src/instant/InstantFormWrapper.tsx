import classnames from 'classnames';
import * as React from 'react';
import { Button } from '../utils/forms/Buttons';
import * as panelStyling from '../utils/panel/Panel.scss';
import * as styles from './Instant.scss';

interface InstantFormProps {
  heading: string | React.ReactNode;
  btnLabel: string;
  btnAction: () => void;
  btnDisabled?: boolean;
}

export class InstantFormWrapper extends React.Component<InstantFormProps> {
  public render() {
    const { heading, btnLabel, btnAction, btnDisabled, children } = this.props;

    return (
      <section className={classnames(styles.panel, panelStyling.panel)}>
        <header className={styles.header}>
          <h1>{heading}</h1>
        </header>
        {
          children
        }
        <footer className={styles.footer}>
          <Button
            size="lg"
            color="greyWhite"
            onClick={btnAction}
            style={{ width: '100%' }}
            disabled={btnDisabled}
          >
            {btnLabel}
          </Button>
        </footer>
      </section>
    );
  }
}
