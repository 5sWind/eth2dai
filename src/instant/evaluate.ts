import { BigNumber } from 'bignumber.js';
import { combineLatest, Observable, of } from 'rxjs';
import { catchError, first, flatMap, map, startWith, switchMap } from 'rxjs/operators';
import { ReadCalls, ReadCalls$ } from '../blockchain/calls/calls';
import { OfferType } from '../exchange/orderbook/orderbook';
import { InstantFormState } from './instantForm';
import { MessageKind, Position } from './validate';

export interface TradeEvaluationState {
  buyAmount?: BigNumber;
  sellAmount?: BigNumber;
  tradeEvaluationStatus: TradeEvaluationStatus;
  tradeEvaluationError?: any;
  bestPrice?: BigNumber;
}

export enum TradeEvaluationStatus {
  unset = 'unset',
  calculating = 'calculating',
  calculated = 'calculated',
  error = 'error',
}

export function evaluateTrade(
  theCalls$: ReadCalls$, previousState: InstantFormState, state: InstantFormState
): Observable<TradeEvaluationState> | undefined {

  if (
    state.kind === OfferType.buy &&
    state.kind === previousState.kind &&
    state.buyAmount && previousState.buyAmount &&
    state.buyAmount.eq(previousState.buyAmount)
    ||
    state.kind === OfferType.sell &&
    state.kind === previousState.kind &&
    state.sellAmount && previousState.sellAmount &&
    state.sellAmount.eq(previousState.sellAmount)
  ) {
    return undefined;
  }

  if (
    !state.kind || state.kind === OfferType.buy && !state.buyAmount
    || state.buyAmount && state.buyAmount.eq(new BigNumber(0))
    || state.kind === OfferType.sell && !state.sellAmount
    || state.sellAmount && state.sellAmount.eq(new BigNumber(0))
  ) {
    return of({
      tradeEvaluationStatus: TradeEvaluationStatus.unset,
      ...state.kind === OfferType.buy ? { sellAmount: undefined } : { buyAmount: undefined }
    });
  }

  return theCalls$.pipe(
    first(),
    switchMap(calls =>
      combineLatest(
        state.kind === OfferType.buy ? evaluateBuy(calls, state) : evaluateSell(calls, state),
        // tslint:disable-next-line:max-line-length
        // This is some suspicious case. This way it works like we had on OD but needs in-depth investigation.
        getBestPrice(calls, state.buyToken, state.sellToken)
      )
    ),
    map(([evaluation, bestPrice]) => ({
      ...evaluation,
      bestPrice,
      tradeEvaluationStatus: TradeEvaluationStatus.calculated,
    })),
    startWith({ ...state, tradeEvaluationStatus: TradeEvaluationStatus.calculating }),
    catchError(err => {
      return of({
        ...(state.kind === OfferType.buy ? { sellAmount: undefined } : { buyAmount: undefined }),
        bestPrice: undefined,
        tradeEvaluationStatus: TradeEvaluationStatus.error,
        tradeEvaluationError: err,
      });
    }),
  );
}

function evaluateBuy(calls: ReadCalls, state: InstantFormState) {

  const { buyToken, sellToken, buyAmount } = state;

  if (!buyToken || !sellToken || !buyAmount) {
    return of(state);
  }

  const errorItem = (error?: Error) => ({
    sellAmount: undefined,
    message: {
      error,
      kind: MessageKind.orderbookTotalExceeded,
      amount: buyAmount,
      side: 'buy',
      token: buyToken,
      placement: Position.TOP,
      priority: 3
    }
  });

  return calls.otcGetPayAmount({
    sellToken,
    buyToken,
    amount: buyAmount,
  }).pipe(
    switchMap(sellAmount => of(sellAmount.isZero() ? errorItem() : { sellAmount })),
    catchError(error => of(errorItem(error))),
  );
}

function evaluateSell(calls: ReadCalls, state: InstantFormState) {

  const { buyToken, sellToken, sellAmount } = state;

  if (!buyToken || !sellToken || !sellAmount) {
    return of(state);
  }

  const errorItem = (error?: Error) => ({
    buyAmount: undefined,
    message: {
      error,
      kind: MessageKind.orderbookTotalExceeded,
      amount: sellAmount,
      side: 'sell',
      token: sellToken,
      placement: Position.TOP,
      priority: 3
    }
  });

  return calls.otcGetBuyAmount({
    sellToken,
    buyToken,
    amount: sellAmount,
  }).pipe(
    switchMap(buyAmount => of(buyAmount.isZero() ? errorItem() : { buyAmount })),
    catchError(error => of(errorItem(error))),
  );
}

function getBestPrice(
  calls: ReadCalls,
  sellToken: string,
  buyToken: string
): Observable<BigNumber> {
  return calls.otcGetBestOffer({ sellToken, buyToken }).pipe(
    flatMap(offerId =>
      calls.otcOffers(offerId).pipe(
        map(([a, _, b]: BigNumber[]) => {
          return (sellToken === 'DAI' || (sellToken === 'WETH' && buyToken !== 'DAI')) ?
            a.div(b) : b.div(a);
        })
      )
    )
  );
}
