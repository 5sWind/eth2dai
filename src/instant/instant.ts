import { BigNumber } from 'bignumber.js';
import { curry } from 'ramda';
import { merge, Observable, Subject } from 'rxjs';
import { map, scan, shareReplay, startWith, switchMap } from 'rxjs/operators';

import { Balances, DustLimits } from '../balances/balances';
import { Calls, Calls$ } from '../blockchain/calls/calls';
import { InstantOrderData } from '../blockchain/calls/instant';
import { tokens } from '../blockchain/config';
// import { TxState, TxStatus } from '../blockchain/transactions';
import { /*Offer, */OfferType, Orderbook } from '../exchange/orderbook/orderbook';
import { TradingPair } from '../exchange/tradingPair/tradingPair';
import { combineAndMerge } from '../utils/combineAndMerge';
import {
  calculateAmount,
  calculateTotal,
  doGasEstimation,
  FormResetChange,
  GasEstimationStatus,
  HasGasEstimation,
  toBalancesChange,
  toDustLimitChange$,
  toEtherPriceUSDChange,
  toGasPriceChange,
  toOrderbookChange$,
} from '../utils/form';

export enum FormStage {
  editing = 'editing',
  readyToProceed = 'readyToProceed',
  waitingForApproval = 'waitingForApproval'
}

interface FormStageChange {
  kind: FormChangeKind.formStageChange;
  stage: FormStage;
}

export enum MessageKind {
  noAllowance = 'noAllowance',
  insufficientAmount = 'insufficientAmount',
  incredibleAmount = 'incredibleAmount',
  dustAmount = 'dustAmount',
  orderbookTotalExceeded = 'orderbookTotalExceeded',
}

export type Message = {
  kind: MessageKind.noAllowance | MessageKind.insufficientAmount
    | MessageKind.incredibleAmount;
  field: string;
  priority: number;
  token: string;
} | {
  kind: MessageKind.dustAmount;
  field: string;
  priority: number;
  token: string;
  amount: BigNumber;
} | {
  kind: MessageKind.orderbookTotalExceeded
  field: string;
  priority: number;
};

export interface InstantFormState extends HasGasEstimation {
  buyToken: string;
  sellToken: string;
  buyTokenDigits: number;
  sellTokenDigits: number;
  buyAmount?: BigNumber;
  sellAmount?: BigNumber;
  kind?: OfferType;
  buyAllowance?: boolean;
  sellAllowance?: boolean;
  messages: Message[];
  stage: FormStage;
  submit: (state: InstantFormState) => void;
  change: (change: ManualChange) => void;
  orderbook?: Orderbook;
  dustLimitQuote?: BigNumber;
  dustLimitBase?: BigNumber;
  balances?: Balances;
}

export enum FormChangeKind {
  buyAmountFieldChange = 'buyAmountFieldChange',
  sellAmountFieldChange = 'sellAmountFieldChange',
  pairChange = 'pairChange',
  formStageChange = 'stage',
  gasPriceChange = 'gasPrice',
  etherPriceUSDChange = 'etherPriceUSDChange',
  orderbookChange = 'orderbook',
  sellAllowanceChange = 'sellAllowance',
  buyAllowanceChange = 'buyAllowance',
  balancesChange = 'balancesChange',
  dustLimitChange = 'dustLimitChange',
}
export interface BuyAmountChange {
  kind: FormChangeKind.buyAmountFieldChange;
  value?: BigNumber;
}
export interface SellAmountChange {
  kind: FormChangeKind.sellAmountFieldChange;
  value?: BigNumber;
}
export interface PairChange {
  kind: FormChangeKind.pairChange;
  buyToken: string;
  sellToken: string;
}
export interface GasPriceChange {
  kind: FormChangeKind.gasPriceChange;
  value: BigNumber;
}
export interface EtherPriceUSDChange {
  kind: FormChangeKind.etherPriceUSDChange;
  value: BigNumber;
}
export interface OrderbookChange {
  kind: FormChangeKind.orderbookChange;
  orderbook: Orderbook;
}
export interface AllowanceChange {
  kind: FormChangeKind.buyAllowanceChange | FormChangeKind.sellAllowanceChange;
  allowance: boolean;
}
export interface BalancesChange {
  kind: FormChangeKind.balancesChange;
  balances: Balances;
}
export interface DustLimitChange {
  kind: FormChangeKind.dustLimitChange;
  dustLimitBase: BigNumber;
  dustLimitQuote: BigNumber;
}
function toAllowanceChange$(
  kind: FormChangeKind.buyAllowanceChange | FormChangeKind.sellAllowanceChange,
  token: string,
  theAllowance$: (token: string) => Observable<boolean>): Observable<AllowanceChange> {
  return theAllowance$(token).pipe(
    map((allowance: boolean) => ({ kind, allowance } as AllowanceChange))
  );
}

export type ManualChange =
  BuyAmountChange |
  SellAmountChange |
  PairChange;

export type EnvironmentChange =
  GasPriceChange |
  EtherPriceUSDChange |
  AllowanceChange |
  OrderbookChange |
  BalancesChange |
  DustLimitChange;

export type StageChange =
  FormResetChange |
  FormStageChange;

export type InstantFormChange = ManualChange | EnvironmentChange | StageChange;

function instantOrderData(state: InstantFormState): InstantOrderData {
  return {
    buyAmount: state.buyAmount as BigNumber,
    buyToken: state.buyToken,
    sellAmount: state.sellAmount as BigNumber,
    sellToken: state.sellToken,
  };
}

function applyChange(state: InstantFormState, change: InstantFormChange): InstantFormState {
  switch (change.kind) {
    case FormChangeKind.pairChange:
      return {
        ...state,
        buyToken: change.buyToken,
        sellToken: change.sellToken,
        buyAmount: undefined,
        sellAmount: undefined,
      };
    case FormChangeKind.sellAmountFieldChange:
      if (!state.orderbook) {
        return state;
      }
      return {
        ...state,
        kind: OfferType.sell,
        sellAmount: change.value,
        buyAmount: calculateTotal(change.value, state.orderbook.buy),
      };
    case FormChangeKind.buyAmountFieldChange:
      if (!state.orderbook) {
        return state;
      }
      return {
        ...state,
        kind: OfferType.buy,
        buyAmount: change.value,
        sellAmount: calculateAmount(change.value, state.orderbook.buy),
      };
    case FormChangeKind.gasPriceChange:
      return {
        ...state,
        gasPrice: change.value,
        gasEstimationStatus: GasEstimationStatus.unset
      };
    case FormChangeKind.etherPriceUSDChange:
      return {
        ...state,
        etherPriceUsd: change.value,
        gasEstimationStatus: GasEstimationStatus.unset
      };
    case FormChangeKind.orderbookChange:
      return {
        ...state,
        orderbook: change.orderbook
      };
    case FormChangeKind.buyAllowanceChange:
      return { ...state, buyAllowance: change.allowance };
    case FormChangeKind.sellAllowanceChange:
      return { ...state, sellAllowance: change.allowance };
    case FormChangeKind.balancesChange:
      return {
        ...state,
        balances: change.balances,
        gasEstimationStatus: GasEstimationStatus.unset
      };
    case FormChangeKind.dustLimitChange:
      return {
        ...state,
        dustLimitBase: change.dustLimitBase,
        dustLimitQuote: change.dustLimitQuote
      };
  }
  return state;
}

function addGasEstimation(theCalls$: Calls$,
                          state: InstantFormState): Observable<InstantFormState> {
  return doGasEstimation(theCalls$, state, (calls: Calls) =>
    state.messages.length !== 0 ||
    !state.buyAmount ||
    !state.sellAmount ?
      undefined :
      calls.instantOrderEstimateGas(instantOrderData(state))
  );
}

function preValidate(state: InstantFormState): InstantFormState {
  if (state.stage !== FormStage.editing) {
    return state;
  }

  const messages: Message[] = [];

  return {
    ...state,
    messages,
    gasEstimationStatus: GasEstimationStatus.unset,
  } as InstantFormState;
}

function postValidate(state: InstantFormState): InstantFormState {
  if (state.stage !== FormStage.editing) {
    return state;
  }
  if (state.gasEstimationStatus === GasEstimationStatus.calculated &&
    state.buyAmount && state.sellAmount &&
    state.messages.length === 0) {
    return { ...state, stage: FormStage.readyToProceed };
  }
  return { ...state, stage: FormStage.editing };
}

export function createFormController$(
  params: {
    gasPrice$: Observable<BigNumber>;
    allowance$: (token: string) => Observable<boolean>;
    balances$: Observable<Balances>;
    dustLimits$: Observable<DustLimits>;
    orderbook$: Observable<Orderbook>,
    calls$: Calls$;
    etherPriceUsd$: Observable<BigNumber>
  },
  tradingPair: TradingPair
): Observable<InstantFormState> {

  const manualChange$ = new Subject<ManualChange>();

  const environmentChange$ = combineAndMerge(
    toGasPriceChange(params.gasPrice$),
    toEtherPriceUSDChange(params.etherPriceUsd$),
    toOrderbookChange$(params.orderbook$),
    toDustLimitChange$(params.dustLimits$, tradingPair.base, tradingPair.quote),
    toAllowanceChange$(FormChangeKind.buyAllowanceChange, tradingPair.base, params.allowance$),
    toAllowanceChange$(FormChangeKind.sellAllowanceChange, tradingPair.quote, params.allowance$),
    toBalancesChange(params.balances$),
  );

  const initialState: InstantFormState = {
    submit: null as any,
    change: manualChange$.next.bind(manualChange$),
    buyToken: tradingPair.quote,
    sellToken: tradingPair.base,
    buyTokenDigits: tokens[tradingPair.quote].digits,
    sellTokenDigits: tokens[tradingPair.base].digits,
    buyAmount: undefined,
    sellAmount: undefined,
    kind: undefined,
    buyAllowance: undefined,
    sellAllowance: undefined,
    stage: FormStage.editing,
    messages: [],
    gasEstimationStatus: GasEstimationStatus.unset,
  };

  return merge(
    manualChange$,
    environmentChange$,
  ).pipe(
    scan(applyChange, initialState),
    map(preValidate),
    switchMap(curry(addGasEstimation)(params.calls$)),
    map(postValidate),
    startWith(initialState),
    shareReplay(1),
  );
}
