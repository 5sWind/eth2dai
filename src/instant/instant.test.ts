import { setupFakeWeb3ForTesting } from '../blockchain/web3';

setupFakeWeb3ForTesting();

import { BigNumber } from 'bignumber.js';
import jestEach from 'jest-each';
import { omit } from 'lodash';
import { of } from 'rxjs';
import { first } from 'rxjs/operators';

import { Calls$ } from '../blockchain/calls/calls';
import { TxState, TxStatus } from '../blockchain/transactions';
import { createFakeOrderbook, emptyOrderBook } from '../exchange/depthChart/fakeOrderBook';
import { unpack } from '../utils/testHelpers';
import { createFormController$, FormChangeKind } from './instant';

function snapshotify(object: any): any {
  return omit(object, 'change', 'submit');
}

const tradingPair = { base: 'WETH', quote: 'DAI' };

const defaultCalls = {
  instantOrder: null as any,
  instantOrderEstimateGas: () => of(40),
  offerMake: null as any,
  offerMakeEstimateGas: () => of(20),
  offerMakeDirect: null as any,
  offerMakeDirectEstimateGas: () => of(30),
  cancelOffer: null as any,
  cancelOfferEstimateGas: null as any,
  setupMTProxy: null as any,
  setupMTProxyEstimateGas: null as any,
  approve: null as any,
  disapprove: null as any,
  approveWallet: null as any,
  disapproveWallet: null as any,
  wrap: null as any,
  wrapEstimateGas: null as any,
  unwrap: null as any,
  unwrapEstimateGas: null as any,
};

const defParams = {
  gasPrice$: of(new BigNumber(0.01)),
  etherPriceUsd$: of(new BigNumber(1)),
  allowance$: () => of(true),
  balances$: of({ DAI: new BigNumber(1000), WETH: new BigNumber(10), ETH: new BigNumber(5) }),
  dustLimits$: of({ DAI: new BigNumber(0.1), WETH: new BigNumber(0.1) }),
  orderbook$: of(emptyOrderBook),
  calls$: of(defaultCalls) as Calls$,
};

const controllerWithFakeOrderBook = (buys: any = [], sells: any = [], overrides: {} = {}) => {
  const orderbook = createFakeOrderbook(buys, sells);
  orderbook.buy.forEach((v, i) => v.offerId = new BigNumber(i + 1));
  orderbook.sell.forEach((v, i) => v.offerId = new BigNumber(i + 1));
  orderbook.buy.forEach((v) => v.timestamp = new Date('2019-02-19'));
  orderbook.sell.forEach((v) => v.timestamp = new Date('2019-02-19'));
  return createFormController$(
    {
      ...defParams,
      ...overrides,
      orderbook$: of(orderbook),
    },
    tradingPair
  );
};

const fakeOrderBook = [
  [
    { price: 180, amount: 10 },
  ],
  [
    { price: 200, amount: 10 },
  ]
];

test('initial state', done => {
  const controller = createFormController$(defParams, tradingPair);
  controller.subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
    done();
  });
});

test('change pair', done => {
  const controller = createFormController$(defParams, tradingPair);
  const { change } = unpack(controller);

  change({ kind: FormChangeKind.pairChange, buyToken: 'WETH', sellToken: 'DAI' });

  controller.subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
    done();
  });
});

jestEach([
  ['sell', (change: any) => {
    change({ kind: FormChangeKind.sellAmountFieldChange, value: new BigNumber(1) });
  }],
  ['buy', (change: any) => {
    change({ kind: FormChangeKind.buyAmountFieldChange, value: new BigNumber(90) });
  }],
  ['pay eth', (change: any) => {
    change({ kind: FormChangeKind.pairChange, buyToken: 'DAI', sellToken: 'ETH' });
    change({ kind: FormChangeKind.buyAmountFieldChange, value: new BigNumber(90) });
  }],
  ['buy eth', (change: any) => {
    change({ kind: FormChangeKind.pairChange, buyToken: 'ETH', sellToken: 'DAI' });
    change({ kind: FormChangeKind.buyAmountFieldChange, value: new BigNumber(2) });
  }],
]).test('transaction - %s', (_test, perform, done) => {
  const instantOrderMock = jest.fn(
    () => of({
      status: TxStatus.WaitingForApproval
    } as TxState),
  ).mockName('instantOrder');

  const controller = controllerWithFakeOrderBook(fakeOrderBook[0], fakeOrderBook[1], {
    calls$: of({
      ...defaultCalls,
      instantOrder: instantOrderMock,
    })
  });

  perform(unpack(controller).change);

  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
  });

  unpack(controller).submit(unpack(controller));
  expect(instantOrderMock.mock.calls).toMatchSnapshot();

  controller.subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
    done();
  });
});

test('complex scenario', done => {
  const controller = controllerWithFakeOrderBook(...fakeOrderBook);
  const { change } = unpack(controller);

  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
    done();
  });

  change({ kind: FormChangeKind.sellAmountFieldChange, value: new BigNumber(2) });
  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
  });

  change({ kind: FormChangeKind.buyAmountFieldChange, value: new BigNumber(90) });
  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
  });

  change({ kind: FormChangeKind.pairChange, buyToken: 'WETH', sellToken: 'DAI' });
  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
  });

  change({ kind: FormChangeKind.buyAmountFieldChange, value: new BigNumber(2) });
  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
  });

  change({ kind: FormChangeKind.sellAmountFieldChange, value: new BigNumber(100) });
  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
  });

  change({ kind: FormChangeKind.formResetChange });
  controller.pipe(first()).subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
  });
});

jestEach([
  ['insufficient balance', (change: any) => {
    change({ kind: FormChangeKind.pairChange, buyToken: 'WETH', sellToken: 'DAI' });
    change({ kind: FormChangeKind.sellAmountFieldChange, value: new BigNumber(1100) });
  }],
  ['dust limit', (change: any) => {
    change({ kind: FormChangeKind.sellAmountFieldChange, value: new BigNumber(0.001) });
  }],
  ['incredible amount base', (change: any) => {
    change({ kind: FormChangeKind.sellAmountFieldChange, value: new BigNumber(1000000000000001) });
  }],
  ['incredible amount quote', (change: any) => {
    change({ kind: FormChangeKind.pairChange, buyToken: 'WETH', sellToken: 'DAI' });
    change({ kind: FormChangeKind.sellAmountFieldChange, value: new BigNumber(1000000000000001) });
  }],
  ['orderbook exceeded quote', (change: any) => {
    change({ kind: FormChangeKind.buyAmountFieldChange, value: new BigNumber(2001) });
  }],
  ['orderbook exceeded base', (change: any) => {
    change({ kind: FormChangeKind.pairChange, buyToken: 'WETH', sellToken: 'DAI' });
    change({ kind: FormChangeKind.buyAmountFieldChange, value: new BigNumber(11) });
  }],
]).test('validation - %s', (_test, perform, done) => {
  const controller = controllerWithFakeOrderBook(...fakeOrderBook);
  perform(unpack(controller).change);
  controller.subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
    done();
  });
});
