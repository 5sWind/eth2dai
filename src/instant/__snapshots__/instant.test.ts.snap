// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buy a bit 1`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": "90",
  "buyToken": "DAI",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimation": 40,
  "gasEstimationEth": "4e-19",
  "gasEstimationStatus": "calculated",
  "gasEstimationUsd": "4e-19",
  "gasPrice": "0.01",
  "kind": "buy",
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "180",
        "quoteAmount": "1800",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "sell": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "200",
        "quoteAmount": "2000",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "spread": "20",
  },
  "sellAllowance": true,
  "sellAmount": "0.5",
  "sellToken": "WETH",
  "sellTokenDigits": 5,
  "stage": "readyToProceed",
  "submit": null,
}
`;

exports[`change pair 1`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": undefined,
  "buyToken": "WETH",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimationStatus": "unset",
  "gasPrice": "0.01",
  "kind": undefined,
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [],
    "sell": Array [],
    "spread": undefined,
  },
  "sellAllowance": true,
  "sellAmount": undefined,
  "sellToken": "DAI",
  "sellTokenDigits": 5,
  "stage": "editing",
  "submit": null,
}
`;

exports[`complex scenario 1`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": undefined,
  "buyToken": "DAI",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimationStatus": "unset",
  "gasPrice": "0.01",
  "kind": undefined,
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "180",
        "quoteAmount": "1800",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "sell": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "200",
        "quoteAmount": "2000",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "spread": "20",
  },
  "sellAllowance": true,
  "sellAmount": undefined,
  "sellToken": "WETH",
  "sellTokenDigits": 5,
  "stage": "editing",
  "submit": null,
}
`;

exports[`complex scenario 2`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": "180",
  "buyToken": "DAI",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimation": 40,
  "gasEstimationEth": "4e-19",
  "gasEstimationStatus": "calculated",
  "gasEstimationUsd": "4e-19",
  "gasPrice": "0.01",
  "kind": "buy",
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "180",
        "quoteAmount": "1800",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "sell": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "200",
        "quoteAmount": "2000",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "spread": "20",
  },
  "sellAllowance": true,
  "sellAmount": "1",
  "sellToken": "WETH",
  "sellTokenDigits": 5,
  "stage": "readyToProceed",
  "submit": null,
}
`;

exports[`complex scenario 3`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": "360",
  "buyToken": "DAI",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimation": 40,
  "gasEstimationEth": "4e-19",
  "gasEstimationStatus": "calculated",
  "gasEstimationUsd": "4e-19",
  "gasPrice": "0.01",
  "kind": "sell",
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "180",
        "quoteAmount": "1800",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "sell": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "200",
        "quoteAmount": "2000",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "spread": "20",
  },
  "sellAllowance": true,
  "sellAmount": "2",
  "sellToken": "WETH",
  "sellTokenDigits": 5,
  "stage": "readyToProceed",
  "submit": null,
}
`;

exports[`complex scenario 4`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": undefined,
  "buyToken": "WETH",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimationStatus": "unset",
  "gasPrice": "0.01",
  "kind": "sell",
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "180",
        "quoteAmount": "1800",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "sell": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "200",
        "quoteAmount": "2000",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "spread": "20",
  },
  "sellAllowance": true,
  "sellAmount": undefined,
  "sellToken": "DAI",
  "sellTokenDigits": 5,
  "stage": "editing",
  "submit": null,
}
`;

exports[`initial state 1`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": undefined,
  "buyToken": "DAI",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimationStatus": "unset",
  "gasPrice": "0.01",
  "kind": undefined,
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [],
    "sell": Array [],
    "spread": undefined,
  },
  "sellAllowance": true,
  "sellAmount": undefined,
  "sellToken": "WETH",
  "sellTokenDigits": 5,
  "stage": "editing",
  "submit": null,
}
`;

exports[`sell a bit 1`] = `
Object {
  "balances": Object {
    "DAI": "10",
    "WETH": "10",
  },
  "buyAllowance": true,
  "buyAmount": "180",
  "buyToken": "DAI",
  "buyTokenDigits": 2,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimation": 40,
  "gasEstimationEth": "4e-19",
  "gasEstimationStatus": "calculated",
  "gasEstimationUsd": "4e-19",
  "gasPrice": "0.01",
  "kind": "sell",
  "messages": Array [],
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "180",
        "quoteAmount": "1800",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "sell": Array [
      Object {
        "baseAmount": "10",
        "baseToken": "",
        "offerId": "1",
        "ownerId": "",
        "price": "200",
        "quoteAmount": "2000",
        "quoteToken": "",
        "timestamp": 2019-02-19T00:00:00.000Z,
        "type": "buy",
      },
    ],
    "spread": "20",
  },
  "sellAllowance": true,
  "sellAmount": "1",
  "sellToken": "WETH",
  "sellTokenDigits": 5,
  "stage": "readyToProceed",
  "submit": null,
}
`;
